addon/-private/wrap-descriptor.js

```diff
@@ -0,0 +1,100 @@
+import { run as runValidator } from './validators';
+
+/**
+ * Transforms a `field` descriptor into a property `getter/setter` that
+ * ensures that the value matches the validators.
+ *
+ * @param {Object} element
+ * @param {Function} validator
+ */
+function wrapFieldDescriptor(
+  { descriptor, initializer, key, placement },
+  validator
+) {
+  const { ...descriptorToApply } = descriptor;
+  delete descriptorToApply.writable;
+
+  let initialized = false;
+  let cachedValue;
+
+  const newElement = {
+    key,
+    placement,
+    kind: 'method',
+    descriptor: {
+      ...descriptorToApply,
+      get() {
+        if (!initialized) {
+          initialized = true;
+          cachedValue = initializer ? initializer.call(this) : undefined;
+        }
+
+        return cachedValue;
+      },
+      set(newValue) {
+        initialized = true;
+
+        runValidator(validator, this.constructor, key, newValue, 'set');
+
+        cachedValue = newValue;
+      }
+    }
+  };
+
+  return newElement;
+}
+
+/**
+ * Wraps a property `getter/setter` such that the value must match the
+ * validator.
+ *
+ * @param {Object} element
+ * @param {Function} validator
+ */
+function wrapMethodDescriptor({ descriptor, key, ...rest }, validator) {
+  const { get, set, ...restOfDescriptor } = descriptor;
+
+  const newElement = {
+    ...rest,
+    key,
+    descriptor: {
+      ...restOfDescriptor,
+      get() {
+        const value = get.call(this);
+
+        runValidator(validator, this.constructor, key, value, 'get');
+
+        return value;
+      },
+      set(newValue) {
+        const setterReturnedValue = set.call(this, newValue);
+
+        runValidator(validator, this.constructor, key, this[key], 'set');
+
+        return setterReturnedValue;
+      }
+    }
+  };
+
+  return newElement;
+}
+
+/**
+ * Transforms a `field` descriptor, wrapping it with a setter that performs
```

Comment:
Note to self: this comment description doesn't match the implementation anymore. Update or remote.

Reply:
Fixed!